backend/regalloc/regalloc_rewrite.ml

```diff
@@ -37,6 +40,54 @@ type direction =
   | Load_after_list of Cfg.basic_instruction_list
   | Store_before_list of Cfg.basic_instruction_list
 
+let coalesce_temp_spills_and_reloads cfg_with_infos new_temporaries =
+  let removed_temporaries = Reg.Tbl.create 128 in
+  let coalesce_temp_reloads_per_block _ block =
+    let var_to_temp_for_all = Reg.Tbl.create 8 in
+    let replacements = Reg.Tbl.create 8 in
+    let last_spill = Reg.Tbl.create 8 in
+    let redundant = ref [] in
+    let replace to_replace replace_with =
+      if not (Reg.same to_replace replace_with)
+      then Reg.Tbl.add replacements to_replace replace_with
+    in
+    let update_info_using_inst inst_cell =
+      let inst = DLL.value inst_cell in
+      match inst.desc with
+      | Op Reload -> (
+        let var = inst.arg.(0) in
+        let temp = inst.res.(0) in
+        match Reg.Tbl.find_opt var_to_temp_for_all var with
+        | None -> Reg.Tbl.add var_to_temp_for_all var temp
+        | Some temp_for_all ->
+          redundant := inst_cell :: !redundant;
+          replace temp temp_for_all)
+      | Op Spill -> (
+        let var = inst.res.(0) in
+        let temp = inst.arg.(0) in
+        (match Reg.Tbl.find_opt last_spill var with
+        | None -> ()
+        | Some prev_inst_cell -> redundant := prev_inst_cell :: !redundant);
+        Reg.Tbl.replace last_spill var inst_cell;
+        match Reg.Tbl.find_opt var_to_temp_for_all var with
+        | None -> Reg.Tbl.add var_to_temp_for_all var temp
+        | Some temp_for_all -> replace temp temp_for_all)
+      | _ -> ()
+    in
+    DLL.iter_cell block.body ~f:update_info_using_inst;
+    List.iter ~f:DLL.delete_curr !redundant;
+    Substitution.apply_block_in_place replacements block;
```

Comment:
I am not claiming it will make a
measurable difference, but we
could maybe have fast path for
the case the substitution ends up
being empty.